### PR TITLE
Parquet benchmark for multi column readers

### PR DESCRIPTION
**Pull request by ClarenceThreepwood**
Test plan - (Please fill in how you tested your changes)

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-hive/src/test/java/com/facebook/presto/hive/parquet/BenchmarkParquetPageSource.java [hunk 1]
```diff
@@ -0,0 +1,511 @@
...
+        }
+
+        private Optional<RowExpression> getFilter(int columnIndex, Type type, float filterRate, boolean filterAllowNull, float selectionRateForNonNull)
+        {
+            if (filterRate == NO_FILTER) {
+                return Optional.empty();
+            }
+
+            if (BOOLEAN.equals(type)) {
+                return Optional.of(new ConstantExpression(true, BOOLEAN));
```
[the remark this answers is not in the document]
Comment (ClarenceThreepwood) [reply presto-hive/src/test/java/com/facebook/presto/hive/parquet/BenchmarkParquetPageSource.java:363]: Good catch. Fixed

File: presto-hive/src/test/java/com/facebook/presto/hive/parquet/BenchmarkParquetPageSource.java [hunk 2]
```diff
@@ -0,0 +1,511 @@
...
+
+            if (INTEGER.equals(type)) {
+                return Optional.of(specialForm(
+                        AND,
+                        BOOLEAN,
+                        call(GREATER_THAN_OR_EQUAL.name(),
+                                functionManager.resolveOperator(GREATER_THAN_OR_EQUAL, fromTypes(type, type)),
+                                BOOLEAN,
+                                field(columnIndex, type),
+                                constant((long) (Integer.MIN_VALUE * (1 - selectionRateForNonNull)), type)),
```
[the remark this answers is not in the document]
Comment (ClarenceThreepwood) [reply presto-hive/src/test/java/com/facebook/presto/hive/parquet/BenchmarkParquetPageSource.java:374]: Correct - I had them reversed

File: presto-hive/src/test/java/com/facebook/presto/hive/parquet/BenchmarkParquetPageSource.java [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+            }
+
+            if (INTEGER.equals(type)) {
+                filter = specialForm(AND,
+                        BOOLEAN,
+                        call(GREATER_THAN_OR_EQUAL.name(),
+                                functionManager.resolveOperator(GREATER_THAN_OR_EQUAL, fromTypes(type, type)),
+                                BOOLEAN,
+                                field(columnIndex, type),
+                                constant((long) (Integer.MIN_VALUE * selectionRateForNonNull), type)),
```
Comment (yingsu00) [inline presto-hive/src/test/java/com/facebook/presto/hive/parquet/BenchmarkParquetPageSource.java:380]: @ClarenceThreepwood It looks good, will you be able to post some sample output row counts for the final output pages for the following cases?
 1. no nulls boolean
 2. partial nulls boolean
 3. no nulls int
 4. partial nulls int
Thank you!

**Review by yingsu00 — COMMENTED** [review]
Mostly looks good. Just would like to see the output row counts to make sure the ratios were enforced corretly.

Also the second commit title is still too long and got truncated. We can use the PR title as the second commit title too.

File: presto-hive/src/test/java/com/facebook/presto/hive/parquet/BenchmarkParquetPageSource.java (hunk 3 above)
Comment (ClarenceThreepwood) [reply presto-hive/src/test/java/com/facebook/presto/hive/parquet/BenchmarkParquetPageSource.java:380]: Total Row Count = 10^6


					      0.1|-1				        0.1|0.5

boolean (no nulls)			8182304				2727065

boolean (partial nulls)		8999746				4500412

int32 (no nulls) 			9000758				4499316

int32 (partial nulls)			9000191				4500421

Comment (yingsu00) [reply presto-hive/src/test/java/com/facebook/presto/hive/parquet/BenchmarkParquetPageSource.java:380]: Thanks @ClarenceThreepwood ! The boolean (no nulls) number for 0.1|0.5 doesn't seem right. Can you take a look? What is the output if there's just one column with filter rates 0.1, 0.5 and 0.9?

Comment (ClarenceThreepwood) [reply presto-hive/src/test/java/com/facebook/presto/hive/parquet/BenchmarkParquetPageSource.java:380]: I believe this is the intention and it is the same as in BenchmarkSelectiveReaders. 
For 0.1|0.5, the filter selectivity for booleans (no nulls) works out to 
c1 = (1-0.1)/(1+0.1) = 0.81
c2 = (1-0.5)/(1+0.5) = 0.33
Final count of output rows = c1*c2 = **0.267**


boolean (no nulls)
filterRate | outputRows
--- | ---
0.1 | 8182304
0.5 | 3334245
0.9 | 526971

Comment (ClarenceThreepwood) [reply presto-hive/src/test/java/com/facebook/presto/hive/parquet/BenchmarkParquetPageSource.java:380]: Updated the boolean generation as discussed. The new numbers are 

Total Row Count = 10^6


|filterRate | 0.1\|-1|  0.1\|0.5|
| --- | --- | --- |
|boolean        (no nulls)                |                  8999659      |                   4497122|
|boolean (partial nulls)     |       8999746       |                4500412|
|int32           (no nulls)           |            9000758	   |                    4499316|
|int32     (partial nulls)	      |         9000191	          |            4500421|


boolean (no nulls)
filterRate | outputRows
--- | ---
0.1 | 8999659
0.5 | 5000426
0.9 | 1000889

**Review by yingsu00 — APPROVED** [review]

**Pull request merged**